File: sdk/go/capabilities.go
package axernsdk

import (
	"context"

	"github.com/cofy-x/axern/sdk/go/internal/nodeclient"
)

type CapabilityStatus struct {
	Ready           bool
	Capabilities    []string
	Providers       []CapabilityProviderStatus
	ProviderSummary CapabilityProviderSummary
}

type CapabilityProviderStatus struct {
	Name         string
	State        string
	Available    bool
	Capabilities []string
	Backend      string
	Reason       string
	Dependencies []CapabilityDependencyStatus
}

type CapabilityDependencyStatus struct {
	Name      string
	Available bool
	Reason    string
}

type CapabilityProviderSummary struct {
	Total       int32
	Available   int32
	Degraded    int32
	Unavailable int32
}

func (s *Sandbox) CapabilityStatus(ctx context.Context) (CapabilityStatus, error) {
	node, err := s.nodeClient()
	if err != nil {
		return CapabilityStatus{}, err
	}
	return node.CapabilityStatus(ctx)
}

func (n *NodeSandboxClient) CapabilityStatus(ctx context.Context) (CapabilityStatus, error) {
	if err := n.validate(); err != nil {
		return CapabilityStatus{}, err
	}
	status, err := n.rpcClient().CapabilityStatus(ctx)
	if err != nil {
		return CapabilityStatus{}, mapRPCError(err, "sandbox capability status", n.allocationID)
	}
	return CapabilityStatus{
		Ready:           status.Ready,
		Capabilities:    append([]string(nil), status.Capabilities...),
		Providers:       sdkCapabilityProviders(status.Providers),
		ProviderSummary: sdkCapabilityProviderSummary(status.ProviderSummary),
	}, nil
}

func sdkCapabilityProviders(items []nodeclient.CapabilityProviderStatus) []CapabilityProviderStatus {
	out := make([]CapabilityProviderStatus, 0, len(items))
	for _, item := range items {
		out = append(out, CapabilityProviderStatus{
			Name:         item.Name,
			State:        item.State,
			Available:    item.Available,
			Capabilities: append([]string(nil), item.Capabilities...),
			Backend:      item.Backend,
			Reason:       item.Reason,
			Dependencies: sdkCapabilityDependencies(item.Dependencies),
		})
	}
	return out
}

func sdkCapabilityDependencies(items []nodeclient.CapabilityDependencyStatus) []CapabilityDependencyStatus {
	out := make([]CapabilityDependencyStatus, 0, len(items))
	for _, item := range items {
		out = append(out, CapabilityDependencyStatus{
			Name:      item.Name,
			Available: item.Available,
			Reason:    item.Reason,
		})
	}
	return out
}

func sdkCapabilityProviderSummary(summary nodeclient.CapabilityProviderSummary) CapabilityProviderSummary {
	return CapabilityProviderSummary{
		Total:       summary.Total,
		Available:   summary.Available,
		Degraded:    summary.Degraded,
		Unavailable: summary.Unavailable,
	}
}

File: sdk/typescript/src/node/capabilities.ts
/**
 * @license
 * Copyright 2026 cofy-x
 * SPDX-License-Identifier: Apache-2.0
 */

import { mapRpcError } from "../errors/index.js";
import { unary } from "../generated/proto.js";
import type { CapabilityStatus, NodeCallOptions } from "../types.js";
import type { NodeClientContext } from "./context.js";

export async function capabilityStatus(ctx: NodeClientContext, options: NodeCallOptions = {}): Promise<CapabilityStatus> {
  try {
    const response = await ctx.withAuthRetry(60, (client) =>
      unary<Record<string, unknown>, Record<string, unknown>>(
        client,
        "CapabilityStatus",
        ctx.authRequest({}),
        options.rpcTimeoutMs,
      ),
    );
    return {
      ready: response.ready === true,
      capabilities: strings(response.capabilities),
      providers: records(response.providers).map((provider) => ({
        name: String(provider.name ?? ""),
        state: String(provider.state ?? ""),
        available: provider.available === true,
        capabilities: strings(provider.capabilities),
        backend: String(provider.backend ?? ""),
        reason: String(provider.reason ?? ""),
        dependencies: records(provider.dependencies).map((dependency) => ({
          name: String(dependency.name ?? ""),
          available: dependency.available === true,
          reason: String(dependency.reason ?? ""),
        })),
      })),
      providerSummary: providerSummary(response.provider_summary),
    };
  } catch (error) {
    throw mapRpcError(error, "sandbox capability status", ctx.allocationId);
  }
}

function providerSummary(input: unknown): CapabilityStatus["providerSummary"] {
  const summary = record(input);
  return {
    total: Number(summary.total ?? 0),
    available: Number(summary.available ?? 0),
    degraded: Number(summary.degraded ?? 0),
    unavailable: Number(summary.unavailable ?? 0),
  };
}

function strings(input: unknown): string[] {
  return Array.isArray(input) ? input.map(String) : [];
}

function records(input: unknown): Record<string, unknown>[] {
  return Array.isArray(input) ? input.map(record) : [];
}

function record(input: unknown): Record<string, unknown> {
  return typeof input === "object" && input !== null ? input as Record<string, unknown> : {};
}
